### PR TITLE
Splitting up some of the Docker build steps

### LINKED ISSUE #8762: the script get stuck in Step 9/24 : RUN cd /app/superset/assets         && npm ci         && npm run build         && rm -rf node_modules  ---> Running in ccea2e24031d
### Expected results

what you expected to happen.

### Actual results

what actually happens.

### Environment

(please complete the following information):

- superset version: `superset version`
- python version: `python --version`
- node.js version: `node -v`
- npm version: `npm -v`

### Checklist

Make sure these boxes are checked before submitting your issue - thank you!

- [ ] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [ ] I have reproduced the issue with at least the latest released version of superset.
- [ ] I have checked the issue tracker for the same issue and I haven't found one similar.

### PR DESCRIPTION
**Pull request by craig-rueda**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Some folks were seeing errors when building the updated Dockerfile due to the fact that we were bundling several NPM commands in a single RUN

Fixes #8762 ??

* Splits up several install commands
* Improves caching behavior by moving superset copy to final stage
* Improves docker-compose init script

### REVIEWERS
@dpgaspar

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -26,24 +29,35 @@ RUN mkdir /app \
             libpq-dev \
         && rm -rf /var/lib/apt/lists/*
 
-COPY ./requirements.txt ./docker/requirements-extra.txt ./setup.py ./MANIFEST.in ./README.md ./app/
-COPY superset /app/superset
-
+# First, we just wanna install requirements, which will allow us to utilize the cache
+# in order to only build iff requirements change
```
Comment (mistercrunch) [inline Dockerfile]: NIT: `iff` typo

File: Dockerfile [hunk 2]
```diff
@@ -26,24 +29,35 @@ RUN mkdir /app \
...
+COPY ./requirements.txt ./docker/requirements-extra.txt ./app/
 RUN cd /app \
-        && pip install -r requirements.txt -r requirements-extra.txt \
-        && pip install -e .
+        && pip install -r requirements.txt -r requirements-extra.txt
```
Comment (mistercrunch) [inline Dockerfile]: NIT: Let's add a comment about getting `-extra` out of the lean build, and add instructions on how to layer up the drivers people may want

File: Dockerfile (hunk 1 above)
Comment (craig-rueda) [reply Dockerfile]: Fixed `iff` -> `if and only if`

File: Dockerfile (hunk 2 above)
Comment (craig-rueda) [reply Dockerfile]: Ended up getting it out. Now there's a new target called `dev` which contains it. Other images will build `lean`

**Review by mistercrunch — APPROVED** [review]

**Pull request merged**